perception/autoware_lidar_frnet/lib/lidar_frnet.cpp

```diff
@@ -0,0 +1,280 @@
+  cuda_utils::clear_async(coors_keys_d.get(), network_params_.num_points_profile.max, stream_);
+  cuda_utils::clear_async(num_points_d.get(), 1, stream_);
+  uint32_t num_unique_coors{0};
+
+  CHECK_CUDA_ERROR(cudaMemsetAsync(
+    proj_idxs_d.get(), 0,
+    preprocessing_params_.interpolation.w * preprocessing_params_.interpolation.h *
+      sizeof(uint32_t),
+    stream_));
+  CHECK_CUDA_ERROR(cudaMemsetAsync(
+    proj_2d_d.get(), -1.0,
+    preprocessing_params_.interpolation.w * preprocessing_params_.interpolation.h * sizeof(float) *
+      4,
+    stream_));
+
+  CHECK_CUDA_ERROR(cudaStreamSynchronize(stream_));
+  CHECK_CUDA_ERROR(preprocess_ptr_->projectPoints_launch(
+    cloud_in_d.get(), input_num_points, num_points_d.get(), points_d_.get(), coors_d_.get(),
+    coors_keys_d.get(), proj_idxs_d.get(), proj_2d_d.get()));
+  CHECK_CUDA_ERROR(preprocess_ptr_->interpolatePoints_launch(
+    proj_idxs_d.get(), proj_2d_d.get(), num_points_d.get(), points_d_.get(), coors_d_.get(),
+    coors_keys_d.get()));
+
+  uint32_t num_points{};
+  CHECK_CUDA_ERROR(cudaStreamSynchronize(stream_));
+  CHECK_CUDA_ERROR(
+    cudaMemcpy(&num_points, num_points_d.get(), sizeof(uint32_t), cudaMemcpyDeviceToHost));
+
+  if (
+    num_points < network_params_.num_points_profile.min ||
+    num_points > network_params_.num_points_profile.max) {
+    RCLCPP_ERROR(
+      logger_,
+      "Number of points (%d) after interpolation does not match the profile. Consider to update "
+      "num_points in config file [min: %ld, opt: %ld, max: %ld].",
+      num_points, network_params_.num_points_profile.min, network_params_.num_points_profile.opt,
+      network_params_.num_points_profile.max);
+    return false;
+  }
+
+  preprocess_ptr_->generateUniqueCoors(
+    num_points, coors_d_.get(), coors_keys_d.get(), num_unique_coors, voxel_coors_d_.get(),
+    inverse_map_d_.get());
+
+  if (
+    num_unique_coors < network_params_.num_unique_coors_profile.min ||
+    num_unique_coors > network_params_.num_unique_coors_profile.max) {
+    RCLCPP_ERROR(
+      logger_,
+      "Number of unique coors (%d) does not match the profile. Consider to update num_unique_coors "
+      "in "
+      "config file [min: %ld, opt: %ld, max: %ld].",
+      num_unique_coors, network_params_.num_unique_coors_profile.min,
+      network_params_.num_unique_coors_profile.opt, network_params_.num_unique_coors_profile.max);
+    return false;
+  }
+
+  network_ptr_->setInputShape("points", {2, {num_points, 4}});
+  network_ptr_->setInputShape("coors", {2, {num_points, 3}});
+  network_ptr_->setInputShape("voxel_coors", {2, {num_unique_coors, 3}});
+  network_ptr_->setInputShape("inverse_map", {1, {num_points}});
+
+  return true;
+}
+
+bool LidarFRNet::inference()
+{
+  network_ptr_->enqueueV3(stream_);
+  return true;
+}
+
+bool LidarFRNet::postprocess(
+  const uint32_t input_num_points, CudaUniquePtr<InputPointType[]> & cloud_in_d,
```

Comment:
Users may not subscribe to all outputs. 
In that case there seems to be unnecessary allocations and device->host copies.
Could we not have persistent buffers to create the outputs and only copy them to the host when we know the users are requesting it?